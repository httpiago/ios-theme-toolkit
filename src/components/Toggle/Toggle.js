import React from 'react'
import PropTypes from 'prop-types'
import { mapClassList } from '../../Utils'

/**
 * Toggle component.
 *
 * @since 0.1.0
 * @see https://httpiago.github.io/ios-theme-toolkit/#/components/Toggle
 */
export default function Toggle({ color, size, className: aditionalClasses, ...rest }) {

  const prefix = 'toggle'
  const classes = mapClassList({
    [`${prefix}--${color}`]: color,
    [`${prefix}--${size}`]: size
  })

  return (
    <label>
      <input
        type="checkbox"
        className={`${prefix} ${classes} ${aditionalClasses}`}
        {...rest}
      />
      <div className="toggle__icon" />
    </label>
  )
}

Toggle.defaultProps = {
  color: 'green',
  size: 'default',
  className: ''
}

Toggle.propTypes = {
  /** Aditional classes. */
  className: PropTypes.string,
  /** The checked state for CONTROLLED components. */
  checked: PropTypes.bool,
  /** The initial state value for UNCONTROLLED components. */
  defaultChecked: PropTypes.bool,
  /** Disable or not the toggle. */
  disabled: PropTypes.bool,
  /** Color of progress bar. */
  color: PropTypes.oneOf([ 'default', 'blue', 'red', 'green', 'yellow', 'gray' ]),
  /** size of toggle */
  size: PropTypes.oneOf([ 'default', 'small', 'big' ]),
  /** Function callback when user change the state */
  onChange: PropTypes.func
}
