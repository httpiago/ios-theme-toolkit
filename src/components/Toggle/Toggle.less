@import '../../customizations.less';

@toggle-selector: .toggle[type=checkbox];

// Esconder o input[type="checkbox"]
@{toggle-selector} {
  -webkit-appearance: none;
  appearance: none;
  visibility: hidden;
  outline: none;
  margin: 0px;
  width: 0px;
  height: 0px;
}

// Default styles
@{toggle-selector} + .toggle__icon {
  background: #d6d6d6;
  border-radius: 1000px;
  width: 52px;
  height: 32px;
  z-index: 0;
  margin: 0;
  padding: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: none;
  position: relative;
  -webkit-box-sizing: border-box;
  box-sizing: border-box;
  display: inline-block;
  vertical-align: middle;
  cursor: pointer;

  // Bolinha dentro do toggle
  &::after {
    background: #fff;
    -webkit-box-shadow: 0 2px 4px rgba(0,0,0,.3);
    box-shadow: 0 2px 4px rgba(0,0,0,.3);
    border-radius: 100px;
    position: absolute;
    content: '';
    height: 27px;
    width: 27px;
    z-index: 2;
    -webkit-transform: translateX(0px);
    transform: translateX(0px);
    -webkit-transition-duration: .3s;
    transition-duration: .3s;
    top: 2px;
    left: 2.5px;
  }
}

// Efeito de foco no toggle
@{toggle-selector}:not(:active):focus + .toggle__icon {
  box-shadow: 0 0 0 3px fade(@gray-color, 10%);
}

// Toggle checked
@{toggle-selector}:checked + .toggle__icon {
  background-color: @green-color;

  // Bolinha dentro do toggle
  &::after {
    -webkit-transform: translateX(20px);
    transform: translateX(20px);
  }
}

// Toggle desabilitado
@{toggle-selector}[disabled] + .toggle__icon {
  opacity: .6;
  pointer-events: none;
  cursor: default;
}

// Toggle pequeno
.toggle--small[type=checkbox] + .toggle__icon {
  width: 42px;
  height: 24px;

  // Bolinha dentro do toggle
  &::after {
    width: 19px;
    height: 19px;
    top: 2px;
  }
}
.toggle--small[type=checkbox]:checked + .toggle__icon::after {
  -webkit-transform: translateX(18px);
  transform: translateX(18px);
}

// Toggle grande
.toggle--big[type=checkbox] + .toggle__icon {
  width: 66px;
  height: 40px;

  // Bolinha dentro do toggle
  &::after {
    width: 32px;
    height: 32px;
    top: 3px;
    left: 3px;
  }
}
.toggle--big[type=checkbox]:checked + .toggle__icon::after {
  -webkit-transform: translateX(27px);
  transform: translateX(27px);
}

// Toggle com cor azul
@{toggle-selector}.toggle--blue:checked + .toggle__icon {
  background-color: @blue-color;
}
// Foco com cor azul
@{toggle-selector}.toggle--blue:checked:not(:active):focus + .toggle__icon {
  box-shadow: 0 0 0 3px fade(@blue-color, 30%);
}

// Toggle com cor vermela
@{toggle-selector}.toggle--red:checked + .toggle__icon {
  background-color: @red-color;
}
// Foco com cor vermelha
@{toggle-selector}.toggle--red:checked:not(:active):focus + .toggle__icon {
  box-shadow: 0 0 0 3px fade(@red-color, 30%);
}

// Toggle com cor amarela
@{toggle-selector}.toggle--yellow:checked + .toggle__icon {
  background-color: @yellow-color;
}
// Foco com cor cinza
@{toggle-selector}.toggle--yellow:checked:not(:active):focus + .toggle__icon {
  box-shadow: 0 0 0 3px fade(@yellow-color, 30%);
}

// Toggle com cor cinza
@{toggle-selector}.toggle--gray:checked + .toggle__icon {
  background-color: @gray-color;
}
// Foco com cor cinza
@{toggle-selector}.toggle--gray:checked:not(:active):focus + .toggle__icon {
  box-shadow: 0 0 0 3px fade(@gray-color, 30%);
}
