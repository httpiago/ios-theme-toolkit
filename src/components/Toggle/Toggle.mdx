---
name: Toggle
route: /components/Toggle
menu: Components
---

import { Playground, PropsTable } from 'docz'
import { Toggle } from '../../index.js'
import '../../../dist/styles.css'

# Toggle

To use this component, simply import it into any file and pass the [modifiers](#/#props) in the properties.

```jsx
import { Toggle } from 'ios-theme-toolkit'

export default () => (
  <label>
    <Toggle checked={true} /> Label text
  </label>
)
```

## Examples

<Playground>
  <label>
    Green (default)
    <Toggle defaultChecked name="toggle-demo" value="wifi" />
  </label>

  <label>
    Blue
    <Toggle color="blue" name="toggle-demo" value="gps" defaultChecked={true} />
  </label>

  <label>
    Red (Disabled)
    <Toggle color="red" name="toggle-demo" value="bluetooth" defaultChecked={true} disabled />
  </label>

  <label>
    Yellow
    <Toggle color="yellow" name="toggle-demo" value="sync" defaultChecked={false} />
  </label>

  <label>
    Gray
    <Toggle color="gray" name="toggle-demo" value="battery" defaultChecked={true} />
  </label>

  <label>
    Small toggle
    <Toggle size="small" name="toggle-demo" value="small" defaultChecked />
  </label>

  <label>
    Big toggle
    <Toggle size="big" name="toggle-demo" value="small" defaultChecked />
  </label>

  <style>{".react-live-preview label { display: block; padding: 6px 0; } .react-live-preview .toggle__icon { margin-right: 6px !important; }"}</style>
</Playground>

## Props

<PropsTable of={ Toggle } />

All the properties which are not listed above will be transferred to the \<input type="checkbox"\> tag, so, It accepts all props which native checkbox support.
