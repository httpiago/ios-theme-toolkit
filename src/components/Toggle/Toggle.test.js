/**
 * TEST FILE
 * Please try writing the tests for all component behaviors,
 * manually call the functions within the class, change the properties...
 *
 * @see https://jestjs.io/docs/en/expect
 * @see https://github.com/airbnb/enzyme/tree/master/docs/api
 */
import React from 'react'
import { shallow } from 'enzyme'
import Toggle from './Toggle'

const COMPONENT = shallow(
  <Toggle checked={false} onChange={() => {}} className="aditionalClass" />
)

describe('Test Toggle component', () => {
  it('should render correctly based on last snapshot', () => {
    expect(COMPONENT).toMatchSnapshot()
  })

  it('should update correctly', () => {
    COMPONENT.setProps({ checked: true })
    
    expect(COMPONENT.html()).toContain('checked')
  })

  it('should has aditional classes', () => {
    expect(COMPONENT.find('.toggle').hasClass('aditionalClass')).toBeTruthy()
  })

  it('should pass aditional props to html element', () => {
    COMPONENT.setProps({ 'data-custom-attr': 'yes' })

    expect(COMPONENT.html()).toContain('data-custom-attr="yes"')
  })

  it('should return a input[type="checkbox"] tag', () => {
    expect(COMPONENT.find('.toggle').type()).toEqual('input')

    expect(COMPONENT.find('.toggle').html()).toContain('type="checkbox"')
  })

})
