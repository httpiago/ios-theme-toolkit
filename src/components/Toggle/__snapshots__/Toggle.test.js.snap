// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Toggle component should render correctly based on last snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Toggle
    checked={false}
    className="aditionalClass"
    color="green"
    onChange={[Function]}
    size="default"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          checked={false}
          className="toggle toggle--green aditionalClass"
          onChange={[Function]}
          type="checkbox"
        />,
        <div
          className="toggle__icon"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "checked": false,
          "className": "toggle toggle--green aditionalClass",
          "onChange": [Function],
          "type": "checkbox",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "toggle__icon",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
    ],
    "type": "label",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            checked={false}
            className="toggle toggle--green aditionalClass"
            onChange={[Function]}
            type="checkbox"
          />,
          <div
            className="toggle__icon"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "checked": false,
            "className": "toggle toggle--green aditionalClass",
            "onChange": [Function],
            "type": "checkbox",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "toggle__icon",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
      ],
      "type": "label",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
